spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,312 @@
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReaderFactory;
+import org.apache.spark.sql.connector.read.streaming.MicroBatchStream;
+import org.apache.spark.sql.connector.read.streaming.Offset;
+import org.apache.spark.sql.execution.streaming.HDFSMetadataLog;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.reflect.ClassTag;
+
+import static org.apache.iceberg.TableProperties.SPLIT_LOOKBACK;
+import static org.apache.iceberg.TableProperties.SPLIT_LOOKBACK_DEFAULT;
+import static org.apache.iceberg.TableProperties.SPLIT_OPEN_FILE_COST;
+import static org.apache.iceberg.TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT;
+import static org.apache.iceberg.TableProperties.SPLIT_SIZE;
+import static org.apache.iceberg.TableProperties.SPLIT_SIZE_DEFAULT;
+
+public class SparkMicroBatchStream implements MicroBatchStream {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkMicroBatchStream.class);
+
+  private final JavaSparkContext sparkContext;
+  private final Table table;
+  private final boolean caseSensitive;
+  private final Schema expectedSchema;
+  private final Long splitSize;
+  private final Integer splitLookback;
+  private final Long splitOpenFileCost;
+  private final boolean localityPreferred;
+  private final InitialOffsetStore initialOffsetStore;
+  private final StreamingOffset initialOffset;
+
+  SparkMicroBatchStream(SparkSession spark,
+                        Table table, boolean caseSensitive, Schema expectedSchema,
+                        CaseInsensitiveStringMap options, String checkpointLocation) {
+    this.sparkContext = JavaSparkContext.fromSparkContext(spark.sparkContext());
+    this.table = table;
+    this.caseSensitive = caseSensitive;
+    this.expectedSchema = expectedSchema;
+    this.localityPreferred = Spark3Util.isLocalityEnabled(table.io(), table.location(), options);
+    this.splitSize = Optional.ofNullable(Spark3Util.propertyAsLong(options, SparkReadOptions.SPLIT_SIZE, null))
+        .orElseGet(() -> PropertyUtil.propertyAsLong(table.properties(), SPLIT_SIZE, SPLIT_SIZE_DEFAULT));
+    this.splitLookback = Optional.ofNullable(Spark3Util.propertyAsInt(options, SparkReadOptions.LOOKBACK, null))
+        .orElseGet(() -> PropertyUtil.propertyAsInt(table.properties(), SPLIT_LOOKBACK, SPLIT_LOOKBACK_DEFAULT));
+    this.splitOpenFileCost = Optional.ofNullable(
+        Spark3Util.propertyAsLong(options, SparkReadOptions.FILE_OPEN_COST, null))
+        .orElseGet(() -> PropertyUtil.propertyAsLong(table.properties(), SPLIT_OPEN_FILE_COST,
+            SPLIT_OPEN_FILE_COST_DEFAULT));
+    this.initialOffsetStore = InitialOffsetStore.getInstance(spark, checkpointLocation);
+    this.initialOffset = getOrWriteInitialOffset(table, initialOffsetStore);
```

Comment:
Summarizing an offline conversation around this:

I was thinking that we could maybe skip `getOrWriteInitialOffset` and instead always compute the initialOffset since this is the one place where we use it. Inside of Spark the initialOffset will be used if we're starting from "scratch" (e.g. no records processed).

I don't think caching initialOffset does anything for us, the only thing which it could do is change the behaviour in `scanAllFiles` but we would probably want to set scanAllFiles to true in the situation where the old initial offset no longer exists and we have only one snapShot anyways.

Reply:
@holdenk - thinking this thru - I don't see a way out of using `getOrWriteInitialOffset`. Here's why:

for ex: a table has snapshots S1, S2 & here are the series of events that will happen:
1 -> start a `readStream` on the table
2 -> at which point the connector should return 
                 - start offset as [S1, 0, scanAllFiles = true]
                 - last offset as [S2, lastFilePosition, scanAllFiles = false]
3 -> `readStream` completes processing this MicroBatch and no events came to the table.
4 -> now lets say that the spark cluster running this `readStream` - died and restarted.
5 -> when it died - also - there is an additional Snapshot S3 - that was written to the iceberg table.
6 -> and Snapshots S1 and S2 are deleted - due to retention policy - as these 2 Snapshots expired.
6 -> Now, when the spark cluster came back up and **re-started** the `readStream` on this iceberg table - we need to infer that 
               - **before S3 - this stream already read some events from the table and hence - we need to set `scanAllFiles` in snapshot S3 to false.**
               - **without storing the fact that - the stream started from S1 - there is just no way to implement this!**
               - if we don't persist where the stream started (the `initialOffset`) - then we will have to assume that table's starting offset - might be the initial offset. In the above case `S3` is clearly not the initialOffset for the stream; as a result, we will end up setting scanAllFiles - to true! - which will result in **replaying all the data from S1 and S2** - which breaks the contract of stream.

pl. let me know if this made sense. I am happy to go on a call and explain or hear any other possible solutions.

Reply:
Based on our offline conversation - @holdenk & I found & agreed that - we will need to store the initial offset. resolving this.